homeassistant/components/zwave_js/__init__.py

```diff
@@ -100,6 +101,24 @@ async def async_on_initialized() -> None:
         # update entity availability
         async_dispatcher_send(hass, f"{DOMAIN}_{entry.entry_id}_connection_state")
 
+        # Check for nodes that no longer exist and remove them
+        stored_devices = device_registry.async_entries_for_config_entry(
+            dev_reg, entry.entry_id
+        )
+        known_devices = [
+            dev_reg.async_get_device({get_device_id(client, node)})
+            for node in client.driver.controller.nodes.values()
+        ]
+
+        # Devices that are in the device registry that are not known by the controller can be removed
+        for device in stored_devices:
+            if device not in known_devices:
+                for entity in entity_registry.async_entries_for_device(
+                    ent_reg, device.id, include_disabled_entities=True
+                ):
+                    ent_reg.async_remove(entity.entity_id)
```

Comment:
Entity entries connected to a device will also be removed if the device is removed.

https://github.com/home-assistant/core/blob/aea8636c7e0224deb3877e6cb67444609635dfed/homeassistant/helpers/entity_registry.py#L288-L303